BigSampler Uniform Sampling fails Serialization

```
Exception in thread "main" java.lang.IllegalArgumentException: unable to serialize DoFnAndMainOutput{doFn=anonymous function keyBy@{SamplerSCollectionFunctions.scala:116}, mainOutputTag=Tag<org.apache.beam.sdk.values.PCollection.<init>:364#6929f09b03d242ca>}
    at org.apache.beam.sdk.util.SerializableUtils.serializeToByteArray(SerializableUtils.java:55)
    at org.apache.beam.runners.core.construction.ParDoTranslation.translateDoFn(ParDoTranslation.java:491)
    at org.apache.beam.runners.dataflow.PrimitiveParDoSingleFactory$PayloadTranslator$1.translateDoFn(PrimitiveParDoSingleFactory.java:165)
    at org.apache.beam.runners.core.construction.ParDoTranslation.payloadForParDoLike(ParDoTranslation.java:625)
    at org.apache.beam.runners.dataflow.PrimitiveParDoSingleFactory$PayloadTranslator.payloadForParDoSingle(PrimitiveParDoSingleFactory.java:161)
    at org.apache.beam.runners.dataflow.PrimitiveParDoSingleFactory$PayloadTranslator.translate(PrimitiveParDoSingleFactory.java:144)
    at org.apache.beam.runners.core.construction.PTransformTranslation$KnownTransformPayloadTranslator.translate(PTransformTranslation.java:315)
    at org.apache.beam.runners.core.construction.PTransformTranslation.toProto(PTransformTranslation.java:128)
    at org.apache.beam.runners.core.construction.SdkComponents.registerPTransform(SdkComponents.java:97)
    at org.apache.beam.runners.core.construction.PipelineTranslation$1.visitPrimitiveTransform(PipelineTranslation.java:72)
    at org.apache.beam.sdk.runners.TransformHierarchy$Node.visit(TransformHierarchy.java:657)
    at org.apache.beam.sdk.runners.TransformHierarchy$Node.visit(TransformHierarchy.java:649)
    at org.apache.beam.sdk.runners.TransformHierarchy$Node.access$600(TransformHierarchy.java:311)
    at org.apache.beam.sdk.runners.TransformHierarchy.visit(TransformHierarchy.java:245)
    at org.apache.beam.sdk.Pipeline.traverseTopologically(Pipeline.java:458)
    at org.apache.beam.runners.core.construction.PipelineTranslation.toProto(PipelineTranslation.java:44)
    at org.apache.beam.runners.dataflow.DataflowPipelineTranslator.translate(DataflowPipelineTranslator.java:163)
    at org.apache.beam.runners.dataflow.DataflowRunner.run(DataflowRunner.java:715)
    at org.apache.beam.runners.dataflow.DataflowRunner.run(DataflowRunner.java:176)
    at org.apache.beam.sdk.Pipeline.run(Pipeline.java:313)
    at org.apache.beam.sdk.Pipeline.run(Pipeline.java:299)
    at com.spotify.scio.ScioContext$$anonfun$close$1.apply(ScioContext.scala:416)
    at com.spotify.scio.ScioContext$$anonfun$close$1.apply(ScioContext.scala:401)
    at com.spotify.scio.ScioContext.requireNotClosed(ScioContext.scala:466)
    at com.spotify.scio.ScioContext.close(ScioContext.scala:401)
    at com.spotify.ratatool.samplers.BigSamplerBigQuery$.sampleBigQueryTable(BigSampler.scala:681)
    at com.spotify.ratatool.samplers.BigSampler$.singleInput(BigSampler.scala:194)
    at com.spotify.ratatool.samplers.BigSampler$.run(BigSampler.scala:231)
    at com.spotify.ratatool.tool.Ratatool$.main(Ratatool.scala:44)
    at com.spotify.ratatool.tool.Ratatool.main(Ratatool.scala)
Caused by: java.io.NotSerializableException: com.google.api.services.bigquery.model.TableSchema
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
    at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
    at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
    at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
    at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
    at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
```

--- a/ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala
+++ b/ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala
@@ -569,7 +569,7 @@ private[samplers] object BigSamplerBigQuery extends BigSampler {
         case "RECORD" =>
           vs.foldLeft(hasher)((hasher, vi) =>
             hashTableRow(
-              vi.asInstanceOf[TableRow],
+              TableRow(vi.asInstanceOf[Map[String, AnyRef]].toList: _*),
               subfields.tail.mkString(BigSampler.fieldSep.toString),
               field.getFields.asScala,
               hasher)
@@ -607,8 +607,9 @@ private[samplers] object BigSamplerBigQuery extends BigSampler {
         case "TIME" => v.toString
         case "DATETIME" => v.toString
         case "RECORD" if fieldStr.nonEmpty =>
+          // v is a LinkedHashMap
           getTableRowField(
-            v.asInstanceOf[TableRow],
+            TableRow(v.asInstanceOf[java.util.Map[String, AnyRef]].asScala.toList: _*),
             subfields.tail.mkString(BigSampler.fieldSep.toString),
             field.getFields.asScala)
         case t => throw new UnsupportedOperationException(
@@ -654,8 +655,10 @@ private[samplers] object BigSamplerBigQuery extends BigSampler {
     import BigQueryIO.Write.CreateDisposition.CREATE_IF_NEEDED
     import BigQueryIO.Write.WriteDisposition.WRITE_EMPTY
 
-    def buildKey(schema: TableSchema)(tr: TableRow): Set[String] = {
-      distributionFields.map(f => schema.get(f).toString).toSet
+    def buildKey(schema: => Seq[TableFieldSchema])(tr: TableRow): Set[String] = {
+      distributionFields.map{ f =>
+        getTableRowField(tr, f, schema).toString
+      }.toSet
     }
 
     val patchedBigQueryService = new PatchedBigQueryServicesImpl()
@@ -674,7 +677,7 @@ private[samplers] object BigSamplerBigQuery extends BigSampler {
       val coll = sc.bigQueryTable(inputTbl)
 
       val sampledCollection = sample(coll, fields, fraction, seed, distribution, distributionFields,
-        precision, hashTableRow, buildKey(schema), schemaFields, sizePerKey)
+        precision, hashTableRow, buildKey(schemaFields), schemaFields, sizePerKey)
 
       val r = sampledCollection
         .saveAsBigQuery(outputTbl, schema, WRITE_EMPTY, CREATE_IF_NEEDED, tableDescription = "")

--- a/ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala
+++ b/ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala
@@ -569,7 +569,7 @@ private[samplers] object BigSamplerBigQuery extends BigSampler {
         case "RECORD" =>
           vs.foldLeft(hasher)((hasher, vi) =>
             hashTableRow(
-              TableRow(vi.asInstanceOf[Map[String, AnyRef]].toList: _*),
+              vi.asInstanceOf[TableRow],
               subfields.tail.mkString(BigSampler.fieldSep.toString),
               field.getFields.asScala,
               hasher)

--- a/ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/util/SamplerSCollectionFunctions.scala
+++ b/ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/util/SamplerSCollectionFunctions.scala
@@ -120,7 +120,7 @@ object SamplerSCollectionFunctions {
     val populationPerKey: SideInput[Double] = totalRecords.withSideInputs(keyCount)
       .map{case (c, sic) => (c * prob)/sic(keyCount)}.toSCollection.asSingletonSideInput
     val probPerKey = keyed.countByKey.withSideInputs(populationPerKey).map {
-      case ((k, c), sic) => (k, sic(populationPerKey)/c) }.toSCollection
+      case ((k, c), sic) => (k, min(sic(populationPerKey)/c, 1.0)) }.toSCollection
     (populationPerKey, probPerKey)
   }
 

--- a/ratatool-sampling/src/main/scala/com/spotify/scio/Random.scala
+++ b/ratatool-sampling/src/main/scala/com/spotify/scio/Random.scala
@@ -40,7 +40,13 @@ object Random {
     def processElement(c: DoFn[((K, V), Double), (K, V)]#ProcessContext): Unit = {
       val ((key, value), fraction) = c.element()
 
-      rngs.putIfAbsent(key, init(fraction))
+      /** Epsilon slop to avoid failure from floating point jitter */
+      require(
+        fraction >= (0.0 - RandomSampler.roundingEpsilon)
+          && fraction <= (1.0 + RandomSampler.roundingEpsilon),
+        s"Sampling fractions must be on interval [0, 1], Key: $key, Fraction: $fraction")
+
+      rngs.putIfAbsent(key, init())
       val count = samples(fraction, rngs.get(key))
       var i = 0
       while (i < count) {
@@ -49,7 +55,7 @@ object Random {
       }
     }
 
-    def init(fraction: Double): R
+    def init(): R
     def samples(fraction: Double, rng: R): Int
     def setSeed(seed: Long): Unit = this.seed = seed
 
@@ -60,12 +66,7 @@ object Random {
 
     // TODO: Is seed properly handled here
     // TODO: is it necessary to setSeed for each instance like Spark does?
-    override def init(fraction: Double): JRandom = {
-      /** Epsilon slop to avoid failure from floating point jitter */
-      require(
-        fraction >= (0.0 - RandomSampler.roundingEpsilon)
-          && fraction <= (1.0 + RandomSampler.roundingEpsilon),
-        s"Sampling fractions must be on interval [0, 1]")
+    override def init(): JRandom = {
       val r = RandomSampler.newDefaultRNG
       if (seed > 0) {
         r.setSeed(seed)

